### PR TITLE
[CU-35d6w57] Store decimals in subsquid

### PR DESCRIPTION
**Pull request by andresmechali**
## Issue
https://app.clickup.com/t/35d6w57

### CODE REVIEW

**Comment by dzmitry-lahoda** [comment]
is title still corect? xcm does not propagates decimals nor allows to hook to convert,  or any other known bridge. also unkown assets do. ot have decimals and statemine assets can be without. so decimals only ui side concern only. and usdt is 6 decimals.

**Comment by andresmechali** [comment]
> is title still corect? xcm does not propagates decimals nor allows to hook to convert, or any other known bridge. also unkown assets do. ot have decimals and statemine assets can be without. so decimals only ui side concern only. and usdt is 6 decimals.

@dzmitry-lahoda This is, as you said, for UI purpose. It's only FE who will use the Subsquid queries that are affected by this, in order to correctly show accumulated values that involve multiple currencies.

File: subsquid/src/dbHelper.ts [hunk 1]
```diff
@@ -200,6 +201,17 @@ export async function storeHistoricalLockedValue(
...
+  const assetDecimals: Record<string, number> = {};
+
+  for (const assetId of Object.keys(amountsLocked)) {
+    const asset = await ctx.store.get(Asset, {
+      where: {
+        id: assetId,
+      },
+    });
+
+    assetDecimals[assetId] = asset?.decimals || 12;
```
Comment (dzmitry-lahoda) [inline subsquid/src/dbHelper.ts:213]: please remove 12 from this line.

Comment (dzmitry-lahoda) [reply subsquid/src/dbHelper.ts:213]: retain asset to undefined if it is such in this code.

File: subsquid/src/model/generated/asset.model.ts [hunk 2]
```diff
@@ -27,6 +27,12 @@ export class Asset {
   @Column_("numeric", {transformer: marshal.bigintTransformer, nullable: false})
   price!: bigint
 
+  /**
+   * Defaults to 12 when not specified
```
Comment (dzmitry-lahoda) [inline subsquid/src/model/generated/asset.model.ts:31]: as was dicussed in slack with other people either no default or (that is mine) make 12 only last step decision with TODO item to remove that. do not provide decimals from data layer if it is not aware of decimals.

Comment (dzmitry-lahoda) [reply subsquid/src/model/generated/asset.model.ts:31]: OR add gh/cu item to fix data layer with no hardcode.

File: subsquid/src/dbHelper.ts (hunk 1 above)
Comment (saadjhk) [reply subsquid/src/dbHelper.ts:213]: @andresmechali we should remove this

File: subsquid/src/model/generated/asset.model.ts (hunk 2 above)
Comment (saadjhk) [reply subsquid/src/model/generated/asset.model.ts:31]: > is title still corect? xcm does not propagates decimals nor allows to hook to convert, or any other known bridge. also unkown assets do. ot have decimals and statemine assets can be without. so decimals only ui side concern only. and usdt is 6 decimals.

@andresmechali yeah i believe you dont need to store normalized values, but one thing keep in mind are pablo transactions, when there are two assets involved (they might have a different decimals) we need to be sure in that case whether we will calculate correctly or not (e.g spot price)

**Review by vimukthi-git — APPROVED** [review]

File: subsquid/src/dbHelper.ts (hunk 1 above)
Comment (andresmechali) [reply subsquid/src/dbHelper.ts:213]: If I remove this we cannot properly normalize accumulations. Meaning, if for some reason we don't know the number of decimals, how can we sum it with the previous value in a way that makes sense?

File: subsquid/src/model/generated/asset.model.ts (hunk 2 above)
Comment (andresmechali) [reply subsquid/src/model/generated/asset.model.ts:31]: The issue is that if I don't normalize values, I cannot properly provide a value that makes sense for queries like `totalValueLocked`.

File: subsquid/src/dbHelper.ts (hunk 1 above)
Comment (andresmechali) [reply subsquid/src/dbHelper.ts:213]: Yes I understand, this is meant to be exclusively for the FE to use, to provide total locked values without FE needing to get the decimals and do it themselves. As discussed today, ignoring any token that doesn't have decimals. If this is still not convenient I can send them the values of each token and they can add them up as pleased.

File: subsquid/src/model/generated/asset.model.ts (hunk 2 above)
Comment (andresmechali) [reply subsquid/src/model/generated/asset.model.ts]: @saadjhk regarding your comment about Pablo, we do treat assets independently so that's not an issue

**Review by easteregg — APPROVED** [review]

**Comment by andresmechali** [comment]
@mergify refresh

**Pull request merged**